### PR TITLE
OPT: Fix batched generation with FLAX

### LINKED ISSUE #20666: Generating with Flax fails when using padding
### System Info

- `transformers` version: 4.25.1
- Platform: Linux-5.10.133+-x86_64-with-glibc2.27
- Python version: 3.8.15
- Huggingface_hub version: 0.11.1
- PyTorch version (GPU?): 1.13.0+cu116 (False)
- Tensorflow version (GPU?): 2.9.2 (False)
- Flax version (CPU?/GPU?/TPU?): 0.6.2 (cpu)
- Jax version: 0.3.25
- JaxLib version: 0.3.25
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@patil-suraj, @patrickvonplaten, @LysandreJik

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Generation with GPT-2 and OPT doesn't work when using padding token.
Specifically, the issues are

1. In GPT-2, using builtin |endoftext| as pad token works with padding, but using customized token <|pad|> doesn't work, the generation only repeats `!`, e.g., `<|pad|> My cat is cute!!!!!!!!!!!!!!!`.
2. In GPT-2, despite pad_token_id is optional in generate() function, one has to provide pad_token_id=tokenizer.pad_token_id otherwise error. 
3. In OPT, even using builtin <pad> token as padding token doesn't work, the generation only repeats `<s>`, e.g., `<pad></s>My cat is cute<s><s><s><s><s><s><s><s><s><s><s><s><s><s>`.

The colab to reproduce the problems 1-3 is [here](https://colab.research.google.com/drive/1pcwTRU3snLjz8wTJx_Z5t4Z7IXnxHnDI?usp=sharing). 

A related issue to problem 2 is #18884. 

### Expected behavior

Expect the generation works with padding. E.g., `<pad></s>My cat is super cute and I love her so much. I love her so much. I love`.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #20666 

OPT numerical masking was using `-inf` where the attention mask was `0`, which in turn caused the hidden states to be `nan` and derail the whole generation. Changing to a common masking value (`-1e9`) fixes the issue. I've also taken the opportunity to re-enable the commented out tests :)

### CODE REVIEW

File: src/transformers/models/opt/modeling_flax_opt.py [hunk 1]
```diff
@@ -245,7 +244,7 @@ def __call__(
             attention_bias = lax.select(
                 attention_mask > 0,
                 jnp.full(attention_mask.shape, 0.0).astype(self.dtype),
-                jnp.full(attention_mask.shape, float("-inf")).astype(self.dtype),
+                jnp.full(attention_mask.shape, -1e9).astype(self.dtype),
```
Comment (gante) [inline src/transformers/models/opt/modeling_flax_opt.py:247]: @patrickvonplaten I've noticed that encoder-decoder models in Flax use `float("-inf")` here. While this is typically not a problem, as most generation cases don't set the decoder attention mask, we might run into the same problem if the user tries setting it. Shall we also make the `float("-inf")` -> `-1e9` change in those models?

File: tests/models/opt/test_modeling_flax_opt.py [hunk 2]
```diff
@@ -364,43 +364,39 @@ def test_jitted_batch_generation(self):
...
-    #     expected_output_sentence = [
-    #         "Hello, my dog is a little bit of a dork.\nI'm a little bit",
-    #         "Today, I<s><s><s><s><s><s><s><s><s><s><s><s>"
-    #         # TODO fix this test in next PR
-    #         # "Today, I was in the middle of a conversation with a friend about the",
-    #     ]
-    #     self.assertListEqual(expected_output_sentence, batch_out_sentence)
-    #     # TODO outputs will be similar, fix in next PR
-    #     self.assertListEqual(batch_out_sentence, [non_padded_sentence, padded_sentence])
+    def test_batch_generation(self):
```
Comment (gante) [inline tests/models/opt/test_modeling_flax_opt.py:367]: (no changes in this test, simply removed the `#` :) )

**Review by sgugger — APPROVED** [review]
LGTM but would like another approval from a Flax user before merging :-)

File: src/transformers/models/opt/modeling_flax_opt.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/opt/modeling_flax_opt.py:247]: I think we should ! I remember other models suffering from the same issue!

File: tests/models/opt/test_modeling_flax_opt.py (hunk 2 above)
Comment (ArthurZucker) [reply tests/models/opt/test_modeling_flax_opt.py:367]: Nice!

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks for fixing this!

File: src/transformers/models/opt/modeling_flax_opt.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/opt/modeling_flax_opt.py:247]: Good with me - we can update Bart and preserve the `# Copied from ...` statement

Comment (gante) [reply src/transformers/models/opt/modeling_flax_opt.py:248]: @sanchit-gandhi done ;) (this updated a few models with the same issue)

Comment (sanchit-gandhi) [reply src/transformers/models/opt/modeling_flax_opt.py:248]: Amazing, ty @gante 🙌
